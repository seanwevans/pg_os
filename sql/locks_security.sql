\set ECHO none
SET client_min_messages TO warning;
DROP EXTENSION IF EXISTS pg_os CASCADE;
DROP ROLE IF EXISTS pg_os_limited;
CREATE ROLE pg_os_limited LOGIN;
CREATE EXTENSION pg_os;
\set ECHO queries
\set VERBOSITY terse

-- Prepare data owned by the extension owner
INSERT INTO users (username)
VALUES ('security_test_user')
RETURNING id AS process_owner \gset

INSERT INTO processes (name, state, priority, owner_user_id, duration)
VALUES ('security_test_process', 'ready', 1, :'process_owner', 1)
RETURNING id AS process_id \gset

INSERT INTO threads (process_id, name, state, priority)
VALUES (:'process_id', 'security_thread', 'ready', 1)
RETURNING id AS thread_id \gset

SET ROLE pg_os_limited;

-- Direct table access should fail for the limited role
\set ON_ERROR_STOP off
INSERT INTO mutexes (name) VALUES ('direct_mutex_attempt');
\set ON_ERROR_STOP on

-- Security-definer helpers should still succeed
SELECT create_mutex('limited_mutex') AS create_mutex_result;
SELECT lock_mutex(:'thread_id', 'limited_mutex') AS lock_mutex_result;
SELECT unlock_mutex(:'thread_id', 'limited_mutex') AS unlock_mutex_result;
SELECT create_semaphore('limited_sem', 1, 1) AS create_semaphore_result;
SELECT acquire_semaphore(:'process_id', 'limited_sem') AS acquire_semaphore_result;
SELECT release_semaphore(:'process_id', 'limited_sem') AS release_semaphore_result;

RESET ROLE;

-- Verify helper effects as the extension owner
SELECT name, locked_by_thread FROM mutexes WHERE name = 'limited_mutex';
SELECT name, count FROM semaphores WHERE name = 'limited_sem';

DROP ROLE pg_os_limited;
