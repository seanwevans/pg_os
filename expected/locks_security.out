\set ECHO none
INSERT INTO mutexes (name) VALUES ('direct_mutex_attempt');
ERROR:  permission denied for table mutexes
SELECT create_mutex('limited_mutex') AS create_mutex_result;
 create_mutex_result 
---------------------
 
(1 row)

SELECT lock_mutex(1, 'limited_mutex') AS lock_mutex_result;
 lock_mutex_result 
-------------------
 
(1 row)

SELECT unlock_mutex(1, 'limited_mutex') AS unlock_mutex_result;
 unlock_mutex_result 
---------------------
 
(1 row)

SELECT create_semaphore('limited_sem', 1, 1) AS create_semaphore_result;
 create_semaphore_result 
-------------------------
 
(1 row)

SELECT acquire_semaphore(1, 'limited_sem') AS acquire_semaphore_result;
 acquire_semaphore_result 
---------------------------
 
(1 row)

SELECT release_semaphore(1, 'limited_sem') AS release_semaphore_result;
 release_semaphore_result 
---------------------------
 
(1 row)

SELECT name, locked_by_thread FROM mutexes WHERE name = 'limited_mutex';
     name      | locked_by_thread 
--------------+-------------------
 limited_mutex | 
(1 row)

SELECT name, count FROM semaphores WHERE name = 'limited_sem';
     name     | count 
-------------+-------
 limited_sem |     1
(1 row)
